File: POTD/Day-07_Count-Distinct-Elements-in-Every-Window/Solution.py
class Solution:

    # ✅ OPTIMAL APPROACH (Sliding Window + Dictionary)
    def countDistinct(self, arr, k):

        freq = {}
        result = []

        # Step 1: First window
        for i in range(k):
            freq[arr[i]] = freq.get(arr[i], 0) + 1

        result.append(len(freq))

        # Step 2: Slide the window
        for i in range(k, len(arr)):

            # Remove outgoing element
            freq[arr[i - k]] -= 1
            if freq[arr[i - k]] == 0:
                del freq[arr[i - k]]

            # Add incoming element
            freq[arr[i]] = freq.get(arr[i], 0) + 1

            result.append(len(freq))

        return result


'''
============ OTHER APPROACHES (COMMENTED) ============

1️⃣ Brute Force (Using Set)
- For each window, create a set
- Time: O(n * k)
- Space: O(k)

def countDistinct(arr, k):
    ans = []
    for i in range(len(arr) - k + 1):
        s = set(arr[i:i+k])
        ans.append(len(s))
    return ans

2️⃣ Full Standalone Runnable Version

if __name__ == "__main__":
    arr = [1, 2, 1, 3, 4, 2, 3]
    k = 4
    sol = Solution()
    print(sol.countDistinct(arr, k))
'''

File: POTD/Day-07_Count-Distinct-Elements-in-Every-Window/Solution.cpp
class Solution {
  public:
    // ✅ OPTIMAL APPROACH (Sliding Window + Hash Map)
    vector<int> countDistinct(vector<int> &arr, int k) {

        unordered_map<int, int> freq;
        vector<int> result;

        // Step 1: Process first window
        for (int i = 0; i < k; i++) {
            freq[arr[i]]++;
        }
        result.push_back(freq.size());

        // Step 2: Slide the window
        for (int i = k; i < arr.size(); i++) {

            // Remove outgoing element
            freq[arr[i - k]]--;
            if (freq[arr[i - k]] == 0) {
                freq.erase(arr[i - k]);
            }

            // Add incoming element
            freq[arr[i]]++;

            result.push_back(freq.size());
        }

        return result;
    }
};

/*
================ OTHER APPROACHES (COMMENTED) ================

1️⃣ Brute Force (Using Set for every window)
- For each window, insert k elements into a set
- Time: O(n * k)
- Space: O(k)

vector<int> countDistinct(vector<int> &arr, int k) {
    vector<int> ans;
    for (int i = 0; i <= arr.size() - k; i++) {
        unordered_set<int> s;
        for (int j = i; j < i + k; j++) {
            s.insert(arr[j]);
        }
        ans.push_back(s.size());
    }
    return ans;
}

2️⃣ Using Ordered Map (Same logic, slower)
- map instead of unordered_map
- Time: O(n log k)

3️⃣ Full Standalone Runnable Version

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

int main() {
    vector<int> arr = {1, 2, 1, 3, 4, 2, 3};
    int k = 4;
    Solution sol;
    vector<int> res = sol.countDistinct(arr, k);
    for (int x : res) cout << x << " ";
    return 0;
}
*/
